signac_dashboard/modules/plot_viewer.py

```diff
@@ -0,0 +1,110 @@
+
+
+class PlotViewer(Module):
+    """Displays a plot associated with the job
+
+    The PlotViewer module can display an interactive plot by using the
+    Plotly JavaScript library. For information on the different accepted
+    parameters for the data and layout, refer to the `Plotly JS documentation
+    <https://plotly.com/javascript/>`_.
+
+    Example:
+
+    .. code-block:: python
+
+        from signac_dashboard.modules import PlotViewer
+
+        def plotly_args_function(project):
+            return [
+                ("Card title",  # if empty, the "name" parameter will be used
+                 # each element on the data list is a different trace
+                 [{
+                    "x": [1, 2, 3, 4, 5],  # x coordinates of the trace
+                    "y": [1, 2, 4, 8, 16]  # y coordinates of the trace
+                 }],
+                 {"margin": { "t": 0 } }  # layout specification for the whole plot
+                )
+            ]
+
+        plot_module = PlotViewer(plotly_args=plotly_args_function, context="ProjectContext")
+
+    :param name: Default name for the card. Ignored if the :code:`plotly_args`
+        callable provides one for each card.
+    :type name: str
+    :param plotly_args: A callable that accepts a job (in the :code:`'JobContext'`)
+        or a project (in the :code:`'ProjectContext'`) and returns an iterable. Each
+        element will constitute a new card and will be composed of a tuple of three
+        elements: the card title, the plotly data and the plotly layout specification.
+    :type plotly_args: callable
+    :param context: Supports :code:`'JobContext'` and :code:`'ProjectContext'`.
+    :type context: str
+    """
+
+    _supported_contexts = {"JobContext", "ProjectContext"}
+
+    def __init__(
+        self,
+        name="Plot Viewer",
+        plotly_args: Callable[
+            [Union[Job, Project]], Iterable[Tuple[str, List[Dict], Dict]]
+        ] = lambda _: list(),
+        context="JobContext",
+        template="cards/plot_viewer.html",
+        **kwargs,
+    ):
+
+        super().__init__(
+            name=name,
+            context=context,
+            template=template,
+            **kwargs,
+        )
+        self.plotly_args = plotly_args
+
+    def get_cards(self, job_or_project):
+        return [
+            {
+                "name": title if title else self.name,
+                "content": render_template(
+                    self.template,
+                    jobid=job_or_project.id,
+                    plotlydata=data,
+                    plotlylayout=layout,
+                ),
+            }
+            for title, data, layout in self.plotly_args(job_or_project)
+        ]
+
+    def register(self, dashboard: Dashboard):
+        # Register routes
+        dashboard.app.route("/module/plot_viewer/<path:filename>")(plot_viewer_asset)
+
+        # Register assets
+        assets = ["js/plot_viewer.js", "js/plotly-2.16.1.min.js"]
```

Comment:
This asset is pretty large and it seems like it changes often (could get outdated, at the relatively slow pace of signac-dashboard development). Can we use the CDN instead?

Note that you'll need to rewrite the git history on this branch to prevent the inclusion of this file in the git history. If you need help with this task or are uncertain about how to do it correctly, let me know and I'll take care of it!
```suggestion
        assets = ["js/plot_viewer.js", "https://cdn.plot.ly/plotly-2.16.1.min.js"]
```

Reply:
Well, I have never done it before but I can give it a try. I'll let you know if I need any help. It should not be a problem to commit your changes first, right? (by the way, is it ok if I batch the smaller ones?)

I'm happy to make this change, the asset is definitely very large, but maybe it's not so popular to have to rely on an external service for this module. Maybe a solution where the user can provide a local copy as a fallback would be useful? I guess it would be a niche situation and not work the effort, but I thought I'd mention it.

Reply:
You can commit GitHub suggestions individually or batched. Your decision!

Reply:
> Maybe a solution where the user can provide a local copy as a fallback would be useful? I guess it would be a niche situation and not work the effort, but I thought I'd mention it.

Yeah, I don't think we are intending for signac-dashboard to be fully usable offline. The core features are baked in, but I don't think it's necessary to include large assets for specific modules like this one in the distributed package. It increases the size of the signac-dashboard repo and downloaded package, and has a high likelihood of going out of date.

Reply:
@javierbg to rewrite the history of the branch perform an "interactive rebase" and remove the commits that added the large file.

The dashboard developers could alternately squash merge this branch to prevent the large file addition to the history.